hashes/src/chacha20poly1305/poly1305.rs

```diff
@@ -0,0 +1,242 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Poly1305 one-time message authenticator from RFC8439.
+//!
+//! Heavily inspired by the ["Donna"](https://github.com/floodyberry/poly1305-donna/blob/master/poly1305-donna-32.h) implementation in C
+//! and Loup Vaillant's [Poly1305 design article](https://loup-vaillant.fr/tutorials/poly1305-design).
+
+/// 2^26 for the 26-bit limbs.
+const BITMASK: u32 = 0x03ffffff;
+/// Number is encoded in five 26-bit limbs.
+const CARRY: u32 = 26;
+
+/// Poly1305 authenticator takes a 32-byte one-time key and a message and produces a 16-byte tag.
+///
+/// 64-bit constant time multiplication and addition implementation.
+#[derive(Debug)]
+pub struct Poly1305 {
+    /// r part of the secret key.
+    r: [u32; 5],
+    /// s part of the secret key.
+    s: [u32; 4],
```

Comment:
Oh, you have secret here directly, so I withdraw my request to make the API same as other hash functions. I think we should just put it into the chcha crate, not here then.

Reply:
Of the existing hashes, I believe the poly1305 interface is probably most similar to SipHash24 (poly1305 is cryptographic)? So I could model it after that?

Reply:
Oh, yes, I've completely forgot about it. Though I don't think there should be the `new` method or even `Default` impl. I also suspect using zero keys for siphash isn't right.

Reply:
Yeah, do it without `new`/`Default`. See #3084

Reply:
@Kixunil and @apoelstra , I am gonna push forward on this now and break this off into its own crate, but any thoughts on `bitcoin-chacha20-poly1305` vs. `chacha20-poly1305`? I feel a little more comfortable with the bitcoin prefix version, but don't have a strong preference and am not sure I have a good feel for when `bitcoin-*` should be applied to a crate (I'll admit I assumed every crate in this org would have it).

Reply:
Oh, I realized if the crate depends on `bitcoin_hashes` anyway (though it's a good question if it really should) then `bitcoin-` makes sense. If it's usable as stanalone without any `bitcoin` stuff then I'd prefer it as a general-purpose crate.

Reply:
Agreed with that -- use the `bitcoin-` prefix iff you need `bitcoin_hashes`.

Reply:
I rolled with sans `bitcoin-` prefix and kept all the AEAD parts together in a new `chacha20-poly1305` crate. The only reason I could think of to use `bitcoin_hashes` was to expose `Poly1305` as a keyed hash function for other parts of the ecosystem. I do not believe there is any demand for this at the moment though, so not really justified.

I have the directory holding the crate as `chacha20_poly1305`, snake_case following Rust conventions, but not sure if that is inline with this repository since the crate `bitcoin-` prefixes are implicit.

Reply:
That makes sense and I like it better this way. I think `-` is preferred in *crate names* in this org though my personal opinion is that `-` shouldn't have been allowed to avoid the confusion.

Reply:
Yea, that definitely would have been ideal. I stuck with the `kebab-case` for the crate name.

I think this patch is in a pretty good spot now though. I am going to try and integrate the AEAD into a HPKE library, so might end it with more thoughts on the interface. But I believe that, along with the performance work for SIMD instructions, can come in some followup work.